File: libs/geometry-py/src/geometry_py/frame3d.py
from __future__ import annotations

import math

from geometry_py.point3d import Point3D
from geometry_py.vector3d import Vector3D


class Frame3D:
    """A right-handed 3D coordinate frame with origin and three orthonormal axes.

    Supports coordinate transformations between world and local spaces via
    both matrix-based and dot-product-based methods.
    """

    __slots__ = ("_origin", "_axis_x", "_axis_y", "_axis_z")

    def __init__(
        self,
        origin: Point3D | None = None,
        axis_x: Vector3D | None = None,
        axis_y: Vector3D | None = None,
        axis_z: Vector3D | None = None,
    ) -> None:
        self._origin = origin if origin is not None else Point3D.origin()

        if axis_x is None and axis_y is None and axis_z is None:
            # World / identity frame
            self._axis_x = Vector3D.unit_x()
            self._axis_y = Vector3D.unit_y()
            self._axis_z = Vector3D.unit_z()
        elif axis_x is not None and axis_y is not None and axis_z is not None:
            # 3-axis constructor: normalize all
            self._axis_x = axis_x.normalized()
            self._axis_y = axis_y.normalized()
            self._axis_z = axis_z.normalized()
        elif axis_x is not None and axis_y is not None and axis_z is None:
            # 2-axis constructor: derive Z = normalize(X x Y)
            self._axis_x = axis_x.normalized()
            self._axis_y = axis_y.normalized()
            self._axis_z = axis_x.cross(axis_y).normalized()
        else:
            raise ValueError("Provide either no axes, X+Y, or X+Y+Z")

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------

    @property
    def origin(self) -> Point3D:
        return self._origin

    @property
    def axis_x(self) -> Vector3D:
        return self._axis_x

    @property
    def axis_y(self) -> Vector3D:
        return self._axis_y

    @property
    def axis_z(self) -> Vector3D:
        return self._axis_z

    def set_origin(self, origin: Point3D) -> None:
        self._origin = origin

    def set_axes(self, axis_x: Vector3D, axis_y: Vector3D, axis_z: Vector3D) -> None:
        self._axis_x = axis_x.normalized()
        self._axis_y = axis_y.normalized()
        self._axis_z = axis_z.normalized()

    # ------------------------------------------------------------------
    # Static factory
    # ------------------------------------------------------------------

    @staticmethod
    def world_frame() -> Frame3D:
        return Frame3D()

    # ------------------------------------------------------------------
    # Matrix-based transformations
    # ------------------------------------------------------------------

    def get_transformation_matrix(self) -> list[float]:
        """Return the 4x4 local-to-world transformation matrix (row-major, 16 elements)."""
        ax, ay, az, o = self._axis_x, self._axis_y, self._axis_z, self._origin
        return [
            ax.x, ay.x, az.x, o.x,
            ax.y, ay.y, az.y, o.y,
            ax.z, ay.z, az.z, o.z,
            0.0,  0.0,  0.0,  1.0,
        ]

    def get_inverse_transformation_matrix(self) -> list[float]:
        """Return the 4x4 world-to-local transformation matrix (row-major, 16 elements)."""
        ax, ay, az = self._axis_x, self._axis_y, self._axis_z
        origin_vec = Vector3D(self._origin.x, self._origin.y, self._origin.z)
        return [
            ax.x, ax.y, ax.z, -ax.dot(origin_vec),
            ay.x, ay.y, ay.z, -ay.dot(origin_vec),
            az.x, az.y, az.z, -az.dot(origin_vec),
            0.0,  0.0,  0.0,  1.0,
        ]

    def world_to_local_matrix(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        if isinstance(obj, Point3D):
            m = self.get_inverse_transformation_matrix()
            lx = m[0] * obj.x + m[1] * obj.y + m[2] * obj.z + m[3]
            ly = m[4] * obj.x + m[5] * obj.y + m[6] * obj.z + m[7]
            lz = m[8] * obj.x + m[9] * obj.y + m[10] * obj.z + m[11]
            return Point3D(lx, ly, lz)
        # Vector: rotation only (no translation)
        ax, ay, az = self._axis_x, self._axis_y, self._axis_z
        lx = ax.x * obj.x + ax.y * obj.y + ax.z * obj.z
        ly = ay.x * obj.x + ay.y * obj.y + ay.z * obj.z
        lz = az.x * obj.x + az.y * obj.y + az.z * obj.z
        return Vector3D(lx, ly, lz)

    def local_to_world_matrix(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        if isinstance(obj, Point3D):
            m = self.get_transformation_matrix()
            wx = m[0] * obj.x + m[1] * obj.y + m[2] * obj.z + m[3]
            wy = m[4] * obj.x + m[5] * obj.y + m[6] * obj.z + m[7]
            wz = m[8] * obj.x + m[9] * obj.y + m[10] * obj.z + m[11]
            return Point3D(wx, wy, wz)
        # Vector: rotation only
        ax, ay, az = self._axis_x, self._axis_y, self._axis_z
        wx = ax.x * obj.x + ay.x * obj.y + az.x * obj.z
        wy = ax.y * obj.x + ay.y * obj.y + az.y * obj.z
        wz = ax.z * obj.x + ay.z * obj.y + az.z * obj.z
        return Vector3D(wx, wy, wz)

    # ------------------------------------------------------------------
    # Dot-product-based transformations
    # ------------------------------------------------------------------

    def world_to_local_dot_product(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        if isinstance(obj, Point3D):
            disp = obj - self._origin  # Vector3D
            return Point3D(disp.dot(self._axis_x), disp.dot(self._axis_y), disp.dot(self._axis_z))
        return Vector3D(obj.dot(self._axis_x), obj.dot(self._axis_y), obj.dot(self._axis_z))

    def local_to_world_dot_product(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        ax, ay, az = self._axis_x, self._axis_y, self._axis_z
        if isinstance(obj, Point3D):
            o = self._origin
            return Point3D(
                o.x + obj.x * ax.x + obj.y * ay.x + obj.z * az.x,
                o.y + obj.x * ax.y + obj.y * ay.y + obj.z * az.y,
                o.z + obj.x * ax.z + obj.y * ay.z + obj.z * az.z,
            )
        return Vector3D(
            obj.x * ax.x + obj.y * ay.x + obj.z * az.x,
            obj.x * ax.y + obj.y * ay.y + obj.z * az.y,
            obj.x * ax.z + obj.y * ay.z + obj.z * az.z,
        )

    # ------------------------------------------------------------------
    # Convenience methods (use dot-product internally)
    # ------------------------------------------------------------------

    def world_to_local(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        return self.world_to_local_dot_product(obj)

    def local_to_world(self, obj: Point3D | Vector3D) -> Point3D | Vector3D:
        return self.local_to_world_dot_product(obj)

    # ------------------------------------------------------------------
    # Frame operations
    # ------------------------------------------------------------------

    def is_orthonormal(self, tolerance: float = 1e-10) -> bool:
        if abs(self._axis_x.length_squared() - 1.0) > tolerance:
            return False
        if abs(self._axis_y.length_squared() - 1.0) > tolerance:
            return False
        if abs(self._axis_z.length_squared() - 1.0) > tolerance:
            return False
        if abs(self._axis_x.dot(self._axis_y)) > tolerance:
            return False
        if abs(self._axis_y.dot(self._axis_z)) > tolerance:
            return False
        if abs(self._axis_z.dot(self._axis_x)) > tolerance:
            return False
        return True

    # ------------------------------------------------------------------
    # String representation
    # ------------------------------------------------------------------

    def __repr__(self) -> str:
        return (
            f"Frame3D{{\n"
            f"  Origin: {self._origin}\n"
            f"  X-Axis: {self._axis_x}\n"
            f"  Y-Axis: {self._axis_y}\n"
            f"  Z-Axis: {self._axis_z}\n"
            f"}}"
        )

    def __str__(self) -> str:
        return self.__repr__()

File: libs/geometry/src/Frame3D.cc
#include "geometry/Frame3D.hh"

#include <cmath>
#include <stdexcept>

namespace geometry {
// ============================================================================
// Constructors
// ============================================================================

Frame3D::Frame3D()
    : m_origin(Point3D::origin()), m_axisX(Vector3D::unitX()), m_axisY(Vector3D::unitY()),
      m_axisZ(Vector3D::unitZ()) {}

Frame3D::Frame3D(const Point3D &origin, const Vector3D &axisX, const Vector3D &axisY,
                 const Vector3D &axisZ)
    : m_origin(origin), m_axisX(axisX.normalized()), m_axisY(axisY.normalized()),
      m_axisZ(axisZ.normalized()) {}

Frame3D::Frame3D(const Point3D &origin, const Vector3D &axisX, const Vector3D &axisY)
    : m_origin(origin), m_axisX(axisX.normalized()), m_axisY(axisY.normalized()),
      m_axisZ(axisX.cross(axisY).normalized()) {}

// ============================================================================
// Accessors
// ============================================================================

void Frame3D::setOrigin(const Point3D &origin) {
    m_origin = origin;
}

void Frame3D::setAxes(const Vector3D &axisX, const Vector3D &axisY, const Vector3D &axisZ) {
    m_axisX = axisX.normalized();
    m_axisY = axisY.normalized();
    m_axisZ = axisZ.normalized();
}

// ============================================================================
// Matrix-Based Transformation Methods
// ============================================================================

/*
 * MATRIX TRANSFORMATION THEORY:
 * ============================
 *
 * A 4x4 homogeneous transformation matrix combines rotation and translation:
 *
 *     | Xx  Yx  Zx  Ox |   | px |   | Xx*px + Yx*py + Zx*pz + Ox |
 * M = | Xy  Yy  Zy  Oy | * | py | = | Xy*px + Yy*py + Zy*pz + Oy |
 *     | Xz  Yz  Zz  Oz |   | pz |   | Xz*px + Yz*py + Zz*pz + Oz |
 *     | 0   0   0   1  |   | 1  |   | 1                          |
 *
 * Where:
 * - (Xx, Xy, Xz) = X-axis of local frame in world coordinates
 * - (Yx, Yy, Yz) = Y-axis of local frame in world coordinates
 * - (Zx, Zy, Zz) = Z-axis of local frame in world coordinates
 * - (Ox, Oy, Oz) = Origin of local frame in world coordinates
 *
 * The inverse (world to local) is efficient for orthonormal matrices:
 * - Rotation inverse = transpose
 * - Translation needs to be rotated by inverse rotation
 */

Point3D Frame3D::worldToLocalMatrix(const Point3D &worldPoint) const {
    // Get the inverse transformation matrix
    const auto invMatrix = getInverseTransformationMatrix();

    // Apply transformation: P_local = M^(-1) * P_world
    // Using homogeneous coordinates (w = 1)
    const double lx = invMatrix[0] * worldPoint.x + invMatrix[1] * worldPoint.y
                    + invMatrix[2] * worldPoint.z + invMatrix[3];
    const double ly = invMatrix[4] * worldPoint.x + invMatrix[5] * worldPoint.y
                    + invMatrix[6] * worldPoint.z + invMatrix[7];
    const double lz = invMatrix[8] * worldPoint.x + invMatrix[9] * worldPoint.y
                    + invMatrix[10] * worldPoint.z + invMatrix[11];

    return {lx, ly, lz};
}

Point3D Frame3D::localToWorldMatrix(const Point3D &localPoint) const {
    // Get the transformation matrix
    const auto matrix = getTransformationMatrix();

    // Apply transformation: P_world = M * P_local
    const double wx =
        matrix[0] * localPoint.x + matrix[1] * localPoint.y + matrix[2] * localPoint.z + matrix[3];
    const double wy =
        matrix[4] * localPoint.x + matrix[5] * localPoint.y + matrix[6] * localPoint.z + matrix[7];
    const double wz = matrix[8] * localPoint.x + matrix[9] * localPoint.y
                    + matrix[10] * localPoint.z + matrix[11];

    return {wx, wy, wz};
}

Vector3D Frame3D::worldToLocalMatrix(const Vector3D &worldVector) const {
    // Vectors only use the rotation part (no translation)
    // For orthonormal frame, inverse rotation = transpose
    const double lx =
        m_axisX.x * worldVector.x + m_axisX.y * worldVector.y + m_axisX.z * worldVector.z;
    const double ly =
        m_axisY.x * worldVector.x + m_axisY.y * worldVector.y + m_axisY.z * worldVector.z;
    const double lz =
        m_axisZ.x * worldVector.x + m_axisZ.y * worldVector.y + m_axisZ.z * worldVector.z;

    return {lx, ly, lz};
}

Vector3D Frame3D::localToWorldMatrix(const Vector3D &localVector) const {
    // V_world = R * V_local (rotation matrix times local vector)
    const double wx =
        m_axisX.x * localVector.x + m_axisY.x * localVector.y + m_axisZ.x * localVector.z;
    const double wy =
        m_axisX.y * localVector.x + m_axisY.y * localVector.y + m_axisZ.y * localVector.z;
    const double wz =
        m_axisX.z * localVector.x + m_axisY.z * localVector.y + m_axisZ.z * localVector.z;

    return {wx, wy, wz};
}

std::array<double, 16> Frame3D::getTransformationMatrix() const {
    /*
     * Row-major 4x4 matrix:
     *
     * | m[0]  m[1]  m[2]  m[3]  |   | Xx  Yx  Zx  Ox |
     * | m[4]  m[5]  m[6]  m[7]  | = | Xy  Yy  Zy  Oy |
     * | m[8]  m[9]  m[10] m[11] |   | Xz  Yz  Zz  Oz |
     * | m[12] m[13] m[14] m[15] |   | 0   0   0   1  |
     */
    return {
        m_axisX.x, m_axisY.x, m_axisZ.x, m_origin.x, // Row 0
        m_axisX.y, m_axisY.y, m_axisZ.y, m_origin.y, // Row 1
        m_axisX.z, m_axisY.z, m_axisZ.z, m_origin.z, // Row 2
        0.0,       0.0,       0.0,       1.0         // Row 3
    };
}

std::array<double, 16> Frame3D::getInverseTransformationMatrix() const {
    /*
     * For orthonormal frame, the inverse is:
     *
     * | Xx  Xy  Xz  -X·O |
     * | Yx  Yy  Yz  -Y·O |
     * | Zx  Zy  Zz  -Z·O |
     * | 0   0   0    1   |
     *
     * The 3x3 rotation part is transposed.
     * The translation part is: -R^T * O = -(axis · origin)
     */
    const Vector3D originVec(m_origin.x, m_origin.y, m_origin.z);

    return {
        m_axisX.x, m_axisX.y, m_axisX.z, -m_axisX.dot(originVec), // Row 0
        m_axisY.x, m_axisY.y, m_axisY.z, -m_axisY.dot(originVec), // Row 1
        m_axisZ.x, m_axisZ.y, m_axisZ.z, -m_axisZ.dot(originVec), // Row 2
        0.0,       0.0,       0.0,       1.0                      // Row 3
    };
}

// ============================================================================
// Dot Product-Based Transformation Methods
// ============================================================================

/*
 * DOT PRODUCT TRANSFORMATION THEORY:
 * ==================================
 *
 * This approach uses the geometric interpretation of dot product:
 *   a · b = |a| * |b| * cos(θ) = projection of a onto b (if b is unit)
 *
 * For orthonormal basis vectors, the dot product gives the component
 * of a vector along each axis directly.
 *
 * WORLD TO LOCAL:
 * ---------------
 * Given point P in world coordinates:
 * 1. Compute displacement from frame origin: V = P - Origin
 * 2. Project V onto each axis:
 *    local.x = V · axisX  (how far along X?)
 *    local.y = V · axisY  (how far along Y?)
 *    local.z = V · axisZ  (how far along Z?)
 *
 * LOCAL TO WORLD:
 * ---------------
 * Given local coordinates (lx, ly, lz):
 *    P_world = Origin + lx*axisX + ly*axisY + lz*axisZ
 *
 * This reconstructs the world position by starting at the frame
 * origin and moving along each axis by the local coordinate amount.
 */

Point3D Frame3D::worldToLocalDotProduct(const Point3D &worldPoint) const {
    // Step 1: Get displacement vector from frame origin to world point
    const Vector3D displacement = worldPoint - m_origin;

    // Step 2: Project displacement onto each axis using dot product
    // For unit vectors: component = displacement · axis
    const double localX = displacement.dot(m_axisX);
    const double localY = displacement.dot(m_axisY);
    const double localZ = displacement.dot(m_axisZ);

    return {localX, localY, localZ};
}

Point3D Frame3D::localToWorldDotProduct(const Point3D &localPoint) const {
    // Start at frame origin
    // Add contribution along each axis: local_coord * axis_vector
    const double worldX =
        m_origin.x + localPoint.x * m_axisX.x + localPoint.y * m_axisY.x + localPoint.z * m_axisZ.x;

    const double worldY =
        m_origin.y + localPoint.x * m_axisX.y + localPoint.y * m_axisY.y + localPoint.z * m_axisZ.y;

    const double worldZ =
        m_origin.z + localPoint.x * m_axisX.z + localPoint.y * m_axisY.z + localPoint.z * m_axisZ.z;

    return {worldX, worldY, worldZ};
}

Vector3D Frame3D::worldToLocalDotProduct(const Vector3D &worldVector) const {
    // For vectors, no translation needed - just project onto axes
    const double localX = worldVector.dot(m_axisX);
    const double localY = worldVector.dot(m_axisY);
    const double localZ = worldVector.dot(m_axisZ);

    return {localX, localY, localZ};
}

Vector3D Frame3D::localToWorldDotProduct(const Vector3D &localVector) const {
    // Reconstruct world vector from local components
    const double worldX =
        localVector.x * m_axisX.x + localVector.y * m_axisY.x + localVector.z * m_axisZ.x;

    const double worldY =
        localVector.x * m_axisX.y + localVector.y * m_axisY.y + localVector.z * m_axisZ.y;

    const double worldZ =
        localVector.x * m_axisX.z + localVector.y * m_axisY.z + localVector.z * m_axisZ.z;

    return {worldX, worldY, worldZ};
}

// ============================================================================
// Convenience Methods
// ============================================================================

Point3D Frame3D::worldToLocal(const Point3D &worldPoint) const {
    // Use dot product method - slightly more efficient (no matrix construction)
    return worldToLocalDotProduct(worldPoint);
}

Point3D Frame3D::localToWorld(const Point3D &localPoint) const {
    return localToWorldDotProduct(localPoint);
}

Vector3D Frame3D::worldToLocal(const Vector3D &worldVector) const {
    return worldToLocalDotProduct(worldVector);
}

Vector3D Frame3D::localToWorld(const Vector3D &localVector) const {
    return localToWorldDotProduct(localVector);
}

// ============================================================================
// Frame Operations
// ============================================================================

bool Frame3D::isOrthonormal(double tolerance) const {
    // Check that axes are unit length
    if (std::abs(m_axisX.lengthSquared() - 1.0) > tolerance)
        return false;
    if (std::abs(m_axisY.lengthSquared() - 1.0) > tolerance)
        return false;
    if (std::abs(m_axisZ.lengthSquared() - 1.0) > tolerance)
        return false;

    // Check that axes are orthogonal (dot products are zero)
    if (std::abs(m_axisX.dot(m_axisY)) > tolerance)
        return false;
    if (std::abs(m_axisY.dot(m_axisZ)) > tolerance)
        return false;
    if (std::abs(m_axisZ.dot(m_axisX)) > tolerance)
        return false;

    return true;
}

Frame3D Frame3D::worldFrame() {
    return {};
}

std::ostream &operator<<(std::ostream &os, const Frame3D &frame) {
    os << "Frame3D{\n"
       << "  Origin: " << frame.m_origin << "\n"
       << "  X-Axis: " << frame.m_axisX << "\n"
       << "  Y-Axis: " << frame.m_axisY << "\n"
       << "  Z-Axis: " << frame.m_axisZ << "\n"
       << "}";
    return os;
}
} // namespace geometry
